lib/pos_mempool.go

```diff
@@ -277,6 +277,88 @@ func (mp *PosMempool) IsRunning() bool {
 	return mp.status == PosMempoolStatusRunning
 }
 
+// OnBlockConnected is an event handler provided by the PoS mempool to handle the blockchain
+// event where a block is connected to the tip of the blockchain. The mempool updates its
+// internal state based on the new block that has been connected.
+//
+// Whenever a block is connected, this event handler removes the block's transactions from
+// the mempool and updates the internal fee estimation to include new block.
+func (mp *PosMempool) OnBlockConnected(blockEvent *BlockEvent) {
+	mp.Lock()
+	defer mp.Unlock()
+
+	if blockEvent == nil || blockEvent.Block == nil || len(blockEvent.Block.Txns) == 0 {
+		return
+	}
+
+	if !mp.IsRunning() {
+		return
+	}
+
+	// Remove all transactions in the block from the mempool.
+	for _, txn := range blockEvent.Block.Txns {
+		txnHash := txn.Hash()
+
+		// This should never happen. We perform a nil check on the txn hash to avoid a panic.
+		if txnHash == nil {
+			continue
+		}
+
+		// Get the transaction from the register. If the txn doesn't exist in the register,
+		// then there's nothing left to do.
+		txn := mp.txnRegister.GetTransaction(txnHash)
+		if txn == nil {
+			continue
+		}
+
+		mp.removeTransactionNoLock(txn, true)
+	}
```

Comment:
I think we also want to call `mp.feeEstimator.AddBlock` with the new block somewhere in here? Or are we managing the past blocks of the fee estimator elsewhere? Also, I need to trace how block connected events are fired in re-orgs. For example, let's say we currently have a chain such as:

A -> B -> C* -> D*, where * denotes uncommitted
and now we reorg to 
A -> B -> E* -> F*

Do we have an OnBlockDisconnected event for C* and D* and an OnBlockConnected event for E* and F*? If so, everything should work as expected. But if that's not the case, then we need to figure out how to add E* to the fee estimator's past blocks cache and remove C* and D* from the past blocks cache.

Reply:
Great call-out. We do need to call `mp.feeEstimator.AddBlock(...)` here, which sorts blocks by block height underneath to internally maintain its sorted list of the last n cached blocks for fee estimation. It's agnostic to the order in which blocks are dinnected, so it's safe as-is.

I have it as a separate task to make sure that the OnBlockConnected and OnBlockDisconnected events will be called from ProcessBlockPoS. It can be a separate PR.